File: occurs.lua
#!/usr/bin/env lua
prog = {
  name = "occurs",
  banner = "occurs 0.85 (27 Sep 2011) by Reuben Thomas <rrt@sc3d.org>",
  purpose = "Count the occurrences of each symbol in a file.",
  notes = "The default symbol type is words (-s \"([[:alpha:]]+)\"); other useful settings\n" ..
    "include:\n\n" ..
    "  non-white-space characters: -s \"[^[:space:]]+\"\n" ..
    "  alphanumerics and underscores: -s \"[[:alnum:]_]+\"\n" ..
    "  XML tags: -s \"<([a-zA-Z_:][a-zA-Z_:.0-9-]*)[[:space:]>]\""
}

require "std"
require "rex_posix"

-- Command-line options
options = {
  Option {{"nocount", "n"}, "don't show the frequencies or total"},
  Option {{"symbol", "s"}, "symbols are given by REGEXP", "Req", "REGEXP"},
}

-- Parse command-line args
os.setlocale ("")
getopt.processArgs ()
local symbolPat = getopt.opt.symbol or "([[:alpha:]]+)"

-- Compile symbol-matching regexp
local ok, pattern = pcall (rex_posix.new, symbolPat)
if not ok then
  die (pattern)
end

-- Process input
local freq = {}
io.processFiles (function (file, number)
                   for line in io.lines () do
                     for s in rex_posix.gmatch (line, pattern) do
                       freq[s] = (freq[s] or 0) + 1
                     end
                   end
               end)

-- Write output
local symbols = 0
for s in pairs (freq) do
  io.writeline (s .. (getopt.opt.nocount and "" or " " .. freq[s]))
  symbols = symbols + 1
end
if not getopt.opt.nocount then
  warn ("total symbols: " .. symbols)
end

File: occurs.rb
#!/usr/bin/env ruby

# By Reuben Thomas, based on Wybo Dekker's "words"

require 'optparse'

NAME = "occurs"

# Command-line arguments
pat, nocount = Regexp.new('\w+'), false
ARGV.options do |opt|
  opt.banner = "#{NAME} - print, and optionally count, words occurring in a list of files"
  opt.on("-n", "--nocount",
         "don't show the frequencies or total") do
    nocount = true end
  opt.on("-s", "--symbol=REGXEP",
         "symbols are given by REGEXP") do |v|
    pat = Regexp.new(v)
  end
  opt.on("-V", "--version",
         "show script version and exit") do
    puts "#{NAME} 0.9 (27 Sep 2011)"
    exit
  end
  opt.parse!
end

# Read input
words = {}
while gets do
  scan(pat).each do |k|
    words[k] ||= 0
    words[k] += 1
  end
end

# Output results
warn "#{words.length} symbols" unless nocount
words.each do |k,v|
  puts nocount ? k : "#{v} #{k}"
end
